## 01. Import Libraries 

In [89]:
import pandas as pd
import numpy as np
import os

## 02. Import Data Sets

In [90]:
# Assign folder path as a string
path = r'D:\02.2022_Instacart Basket Analysis'

In [91]:
# Import products
df_prods =  pd.read_csv(os.path.join(path,'02 Data', 'Original Data','products.csv'),index_col = False)

In [92]:
df_prods.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices'], dtype='object')

In [93]:
df_prods.shape

(49693, 5)

In [94]:
# Import orders
df_ords_wrangled =  pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data','orders_wrangled.csv'),index_col = 0)

D:\Anaconda\lib\site-packages\numpy\lib\arraysetops.py:583: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


In [95]:
df_ords_wrangled.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_time_of_day', 'days_since_prior_order'],
      dtype='object')

In [96]:
df_ords_wrangled.shape

(3421083, 6)

In [97]:
df_ords_wrangled.head()

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## 03. Consistency check

### Mixed-Type Data

In [104]:
# Code for checking mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [105]:
# Correct data type
df_test['mix'] = df_test['mix'].astype('str')

### Missing Value

In [107]:
# Calculate the number of missing data
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [108]:
# Group missing data to df_nan
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [109]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [110]:
# Replace missing values with mean or median value 
df_prods_Revision = df_prods['product_name'].fillna(7, inplace = False) # 7 = any mean value

In [111]:
df_prods_Revision_2 = df_prods['product_name'].fillna(15, inplace = False) # 15 = any median value

In [112]:
df_prods.shape

(49693, 5)

In [113]:
# Filter out missing values (1)
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [114]:
df_prods_clean.shape

(49677, 5)

In [115]:
# Filter out all missing values (2)
df_prods_clean_2 = df_prods.dropna(inplace = False)

In [116]:
df_prods_clean_2.shape

(49677, 5)

In [117]:
# Filter out specific missing values (3)
df_prods_clean_3 = df_prods.dropna(subset = ['product_name'], inplace = False)

In [118]:
df_prods_clean_3.shape

(49677, 5)

### Duplicates

In [120]:
# Assign duplicates in dataframe of df_dups
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [121]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [122]:
df_prods_clean.shape

(49677, 5)

In [123]:
# Clean duplicates and assign remaining values to df_prods_clean_no_dup
df_prods_clean_no_dup = df_prods_clean.drop_duplicates()

In [124]:
df_prods_clean_no_dup.shape

(49672, 5)

## 04. Answer Questions

### (1.) Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further. Tip: Keep an eye on min and max values!

In [126]:
df_prods_clean_no_dup.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [127]:
Max_value_check= df_prods_clean_no_dup.loc[df_prods_clean_no_dup["prices"] > 30]

In [128]:
Max_value_check

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [129]:
df_prods_clean_weirdcheck = df_prods_clean_no_dup.loc[df_prods_clean_no_dup["prices"] < 30]

I filtered the prices starting from 20, which is not far from the 25%, 50%, 75% values and didn't see any abnormality. So, I moved to 30 and check if the prices still make sense. There are only two prices larger than 30, which doesn't seem right, since the prices are too high for cheese and milk. However, I am not sure which prices are they really labeled and I would prefer filter them out of the analysis before I found out the real prices for them, otherwise the statistics would have large deviation.  

In [130]:
df_prods_clean_weirdcheck.shape

(49670, 5)

In [131]:
df_prods_clean_weirdcheck.describe()

product_id      aisle_id  department_id        prices
count  49670.000000  49670.000000   49670.000000  49670.000000
mean   24850.238716     67.761305      11.728770      7.680437
std    14340.931850     38.316060       5.850834      4.199381
min        1.000000      1.000000       1.000000      1.000000
25%    12432.250000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.750000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [132]:
# Exporting
df_prods_clean_weirdcheck.to_csv(os.path.join(path,'02 Data','Prepared Data','products_checked.csv'))

The maximal value for the price is too large and not consistent with min, 25%, 75% value. Therefore, prices should be investigated further.

### (2.) Check for mixed-type data in your df_ords dataframe.

In [134]:
for col in df_ords_wrangled.columns.tolist():
  weird = (df_ords_wrangled[[col]].applymap(type) != df_ords_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords_wrangled[weird]) > 0:
    print (col + ' has mixed type data')
  else:
    print (col + ' is uniform')

order_id is uniform
user_id is uniform
order_number is uniform
orders_day_of_week is uniform
order_time_of_day is uniform
days_since_prior_order is uniform


### (3.) If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

No mixed-type data is found.

### (4.) Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [137]:
df_ords_wrangled.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time_of_day              0
days_since_prior_order    206209
dtype: int64

In [138]:
df_ords_nan = df_ords_wrangled[df_ords_wrangled['days_since_prior_order'].isnull() == True]

In [139]:
df_ords_nan 

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_time_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

There are 206209 missing values in the column "days_since_prior_order". Lost of customers might not place second order, so days since an order prior to their description could not be calculated.

### (5.) Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

My hypothesis is that NaN comes from the customers who have only placed one order. If my hypothesis was true, every NaN should present only when order_number is 1.

In [141]:
df_ords_first_order =  df_ords_wrangled[df_ords_wrangled['order_number'] == 1]

In [142]:
df_ords_first_order.shape

(206209, 6)

The number of first order is the same as the number of missing values, which verifies that all missing values are related to the first order. 

Because deleting missing values would result in the loss of informations from the other columns, it is not a proper method here. Implenmenting mean or median value is also not ideal. Since missing values here result from no entry, mean or median value is nor representative. Therefore, I would create a new column to indicate whether the order is first order or not. 

In [143]:
df_ords_check = df_ords_wrangled

In [144]:
df_ords_check['first_order']= df_ords_wrangled['days_since_prior_order'].isnull() == True

In [145]:
df_ords_check.head()

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  
0                     NaN         True  
1                    15.0        False  
2                    21.0        False  
3                    29.0        False  
4                    28.0        False

In [146]:
# Another method to flag
# df_ords_check['first_order'] = np.where(np.isnan(df_ords_wrangled['days_since_prior_order'].values), True, False)

### (6.) Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [148]:
df_dups = df_ords_wrangled[df_ords_wrangled.duplicated()]

In [149]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_time_of_day, days_since_prior_order, first_order]
Index: []

### (7.) Address the duplicates using an appropriate method. In a markdown cell, explain why you used your method of choice.

In [151]:
df_dups.shape

(0, 7)

No duplicate is found.

In [153]:
df_ords_wrangled.shape

(3421083, 7)

In [154]:
# Exporting
df_ords_wrangled.to_csv(os.path.join(path,'02 Data','Prepared Data','orders_checked.csv'))

products_checked.csv is already saved as in the exercise walkthrough.